core/src/style/pointStyleBuilder.cpp

```diff
@@ -255,6 +274,95 @@ bool PointStyleBuilder::getUVQuad(PointStyle::Parameters& _params, glm::vec4& _q
     return true;
 }
 
+void PointStyleBuilder::labelPointsPlacing(const Line& _line, const glm::vec4& uvsQuad,
+                                           PointStyle::Parameters& params, const DrawRule& _rule) {
+
+    if (_line.size() < 2) { return; }
+
+    auto isOutsideTile = [](const Point& p) {
+        float tolerance = 0.0005;
+        float tile_min = 0.0 + tolerance;
+        float tile_max = 1.0 - tolerance;
+        return ((p.x < tile_min) || (p.x > tile_max) ||
+                (p.y < tile_min) || (p.y > tile_max));
+    };
+
+    auto angleBetween = [](const Point& p, const Point& q) {
+        return RAD_TO_DEG * atan2(q[0] - p[0], q[1] - p[1]);
+    };
+
+    float minLineLength = std::max(params.size.x, params.size.y) * params.placementMinLengthRatio *
+                            m_style.pixelScale() / View::s_pixelsPerTile;
+
+    switch(params.placement) {
+        case LabelProperty::Placement::vertex: {
+            for (size_t i = 0; i < _line.size() - 1; i++) {
+                auto& p = _line[i];
+                auto& q = _line[i+1];
+                if (params.keepTileEdges || !isOutsideTile(p)) {
+                    if (params.autoAngle) {
+                        params.labelOptions.angle = angleBetween(p, q);
+                    }
+                    addLabel(p, uvsQuad, params, _rule);
+                }
+            }
+
+            // Place label on endpoint
+            auto &p = *(_line.rbegin() + 1);
```

Comment:
I don't think we actually need to explicitly add this iteration. The last iteration on the for loop already treats that last case.